lib/app.js

```diff
@@ -30,167 +28,209 @@ function app (opts) {
-    3: cleanInt(ratingHistogram.find('.three .bar-number').text()),
-    2: cleanInt(ratingHistogram.find('.two .bar-number').text()),
-    1: cleanInt(ratingHistogram.find('.one .bar-number').text())
-  };
-  // for other languages
-  const score = parseFloat(ratingBox.find('div.score').text().replace(',', '.')) || 0;
-
-  let video = $('.screenshots span.preview-overlay-container[data-video-url]').attr('data-video-url');
-  if (video) {
-    video = video.split('?')[0];
-  }
-
-  const screenshots = $('.thumbnails .screenshot').toArray().map((elem) => $(elem).attr('src'));
-  const recentChanges = $('.recent-change').toArray().map((elem) => $(elem).text());
-
-  const fields = {
-    title,
-    summary,
-    icon,
-    price,
-    free: price === '0',
-    minInstalls,
-    maxInstalls,
-    score,
-    reviews,
-    developer,
-    developerId,
-    developerEmail,
-    developerWebsite,
-    developerAddress,
-    updated,
-    version,
-    genre: genreText,
-    genreId,
-    familyGenre: familyGenreText,
-    familyGenreId,
-    size,
-    description: descriptionText(description),
-    descriptionHTML: description.html(),
-    histogram,
-    offersIAP,
-    adSupported,
-    androidVersionText,
-    androidVersion,
-    contentRating,
-    screenshots,
-    video,
-    comments,
-    recentChanges,
-    preregister
-  };
-
-  // replace blank values with undefined
-  return R.map((field) => field === '' ? undefined : field, fields);
+/*
+ * Extract the javascript objects returned by the AF_initDataCallback functions
+ * in the script tags of the app detail HTML.
+ */
+function matchScriptData (response) {
+  const scriptRegex = />AF_initDataCallback[\s\S]*?<\/script/g;
+  const keyRegex = /(ds:.*?)'/;
+  const valueRegex = /return ([\s\S]*?)}}\);<\//;
+
+  return response.match(scriptRegex)
+    .reduce((accum, data) => {
+      const keyMatch = data.match(keyRegex);
+      const valueMatch = data.match(valueRegex);
+
+      if (keyMatch && valueMatch) {
+        const key = keyMatch[1];
+        const value = JSON.parse(valueMatch[1]);
+        return R.assoc(key, value, accum);
+      }
+      return accum;
+    }, {});
 }
-function descriptionText (description) {
-  // preserve the line breaks when converting to text
-  const html = '<div>' + description.html().replace(/<\/p>/g, '\n</p>') + '</div>';
-  return cheerio.load(html)('div').text();
-}
-
-function cleanInt (number) {
-  number = number || '0';
-  // removes thousands separator
-  number = number.replace(/\D/g, '');
-  return parseInt(number);
-}
-
-function installNumbers (downloads) {
-  if (!downloads) {
-    return [0, 0];
-  }
-
-  const characters = [' - ', ' et ', '–', '-', '～', ' a '];
-  const char = characters.find((char) => downloads.split(char).length === 2);
-  if (char) return downloads.split(char);
 
-  throw new Error(`Unable to parse min/max downloads: ${downloads}`);
+// TODO may need to do html to text in some of the fields
+
+const MAPPINGS = {
```

Comment:
Why don't we keep this `MAPPINGS` object clean, with only the paths? I think we could use a function to transform the data later, for example using the `R.evolve` with a different object for the fields that need some kind of transformation. The way it's done now makes difficult to express object values, like the `histogram`, which has this as paths: 

```
histogram: {
	1:['ds:10', 0 , 1, 1, 1],
	2:['ds:10', 0 , 1, 2 ,1],
	3:['ds:10', 0 , 1, 3 ,1],
	4:['ds:10', 0 , 1, 4 ,1],
	5:['ds:10', 0 , 1, 5 ,1],  
}
```
I can't simply pass this object without changing the logic that the `extractData` handles the spec when it is an object. This `extractData` wouldn't need to do smth much different when we got an object, simply get the path to it's properties.

So I believe we could do something like: 

```
const tranformData = R.evolve({
  developerId: (devUrl) => devUrl.split('id=')[1],
  score: parseFloat,
  ...
});
...then(matchScriptTag).then(extractData).then(transformData)
```

Reply:
I've just seen your solution for that, it seems good! Thanks 👍